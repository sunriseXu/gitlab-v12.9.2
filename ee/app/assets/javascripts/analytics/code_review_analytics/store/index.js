import Vue from 'vue';
import Vuex from 'vuex';
import * as actions from './actions';
import mutations from './mutations';
import state from './state';

Vue.use(Vuex);

const createStore = () =>
  new Vuex.Store({
    state: state(),
    actions,
    mutations,
  });

export default createStore();
