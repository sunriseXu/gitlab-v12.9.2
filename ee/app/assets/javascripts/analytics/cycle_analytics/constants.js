import { __ } from '~/locale';

export const PROJECTS_PER_PAGE = 50;

export const DEFAULT_DAYS_IN_PAST = 30;

export const EVENTS_LIST_ITEM_LIMIT = 50;

export const EMPTY_STAGE_TEXT = {
  issue: __(
    'The issue stage shows the time it takes from creating an issue to assigning the issue to a milestone, or add the issue to a list on your Issue Board. Begin creating issues to see data for this stage.',
  ),
  plan: __(
    'The planning stage shows the time from the previous step to pushing your first commit. This time will be added automatically once you push your first commit.',
  ),
  code: __(
    'The coding stage shows the time from the first commit to creating the merge request. The data will automatically be added here once you create your first merge request.',
  ),
  test: __(
    'The testing stage shows the time GitLab CI takes to run every pipeline for the related merge request. The data will automatically be added after your first pipeline finishes running.',
  ),
  review: __(
    'The review stage shows the time from creating the merge request to merging it. The data will automatically be added after you merge your first merge request.',
  ),
  staging: __(
    'The staging stage shows the time between merging the MR and deploying code to the production environment. The data will be automatically added once you deploy to production for the first time.',
  ),
  production: __(
    'The production stage shows the total time it takes between creating an issue and deploying the code to production. The data will be automatically added once you have completed the full idea to production cycle.',
  ),
};

export const DEFAULT_STAGE_NAMES = [...Object.keys(EMPTY_STAGE_TEXT), 'total'];

export const TASKS_BY_TYPE_SUBJECT_ISSUE = 'Issue';
export const TASKS_BY_TYPE_SUBJECT_MERGE_REQUEST = 'MergeRequest';
export const TASKS_BY_TYPE_MAX_LABELS = 15;

export const TASKS_BY_TYPE_SUBJECT_FILTER_OPTIONS = {
  [TASKS_BY_TYPE_SUBJECT_ISSUE]: __('Issues'),
  [TASKS_BY_TYPE_SUBJECT_MERGE_REQUEST]: __('Merge Requests'),
};

export const TASKS_BY_TYPE_FILTERS = {
  SUBJECT: 'SUBJECT',
  LABEL: 'LABEL',
};

export const STAGE_ACTIONS = {
  SELECT: 'selectStage',
  EDIT: 'editStage',
  REMOVE: 'removeStage',
  HIDE: 'hideStage',
  CREATE: 'createStage',
  UPDATE: 'updateStage',
};

export const STAGE_NAME = {
  TOTAL: 'total',
  PRODUCTION: 'production',
};
