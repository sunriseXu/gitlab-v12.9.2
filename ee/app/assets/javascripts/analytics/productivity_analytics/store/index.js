import Vue from 'vue';
import Vuex from 'vuex';
import state from './state';
import * as getters from './getters';
import * as actions from './actions';
import mutations from './mutations';
import filters from './modules/filters/index';
import charts from './modules/charts/index';
import table from './modules/table/index';

Vue.use(Vuex);

const createStore = () =>
  new Vuex.Store({
    state: state(),
    getters,
    actions,
    mutations,
    modules: {
      filters,
      charts,
      table,
    },
  });

export default createStore();
