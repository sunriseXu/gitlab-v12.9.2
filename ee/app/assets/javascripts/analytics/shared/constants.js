export const dateFormats = {
  isoDate: 'yyyy-mm-dd',
  defaultDate: 'mmm d, yyyy',
  defaultDateTime: 'mmm d, yyyy h:MMtt',
};

export const scatterChartLineProps = {
  default: {
    type: 'line',
    showSymbol: false,
    // By default zlevel is 2 for all series types.
    // By increasing the zlevel to 3 we make sure that the trendline gets drawn in front of the dots in the chart.
    zlevel: 3,
  },
};

export const LAST_ACTIVITY_AT = 'last_activity_at';

export const DATE_RANGE_LIMIT = 180;

export const OFFSET_DATE_BY_ONE = 1;
