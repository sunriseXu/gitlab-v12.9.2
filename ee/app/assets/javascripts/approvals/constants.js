import { __ } from '~/locale';

export const TYPE_USER = 'user';
export const TYPE_GROUP = 'group';
export const TYPE_HIDDEN_GROUPS = 'hidden_groups';

export const RULE_TYPE_FALLBACK = 'fallback';
export const RULE_TYPE_REGULAR = 'regular';
export const RULE_TYPE_REPORT_APPROVER = 'report_approver';
export const RULE_TYPE_CODE_OWNER = 'code_owner';
export const RULE_TYPE_ANY_APPROVER = 'any_approver';
export const RULE_NAME_ANY_APPROVER = 'All Members';

export const VULNERABILITY_CHECK_NAME = 'Vulnerability-Check';
export const LICENSE_CHECK_NAME = 'License-Check';

export const APPROVAL_RULE_CONFIGS = {
  [VULNERABILITY_CHECK_NAME]: {
    title: __('Vulnerability-Check'),
    popoverText: __(
      'A merge request approval is required when a security report contains a new vulnerability of high, critical, or unknown severity.',
    ),
    documentationText: __('Learn more about Vulnerability-Check'),
  },
  [LICENSE_CHECK_NAME]: {
    title: __('License-Check'),
    popoverText: __(
      'A merge request approval is required when the license compliance report contains a blacklisted license.',
    ),
    documentationText: __('Learn more about License-Check'),
  },
};
