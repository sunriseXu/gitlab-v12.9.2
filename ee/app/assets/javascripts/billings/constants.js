export const TABLE_TYPE_DEFAULT = 'default';
export const TABLE_TYPE_FREE = 'free';
export const TABLE_TYPE_TRIAL = 'trial';
