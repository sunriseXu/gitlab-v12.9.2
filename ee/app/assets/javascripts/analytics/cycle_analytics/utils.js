import { isNumber } from 'underscore';
import dateFormat from 'dateformat';
import { convertObjectPropsToCamelCase } from '~/lib/utils/common_utils';
import { convertToSnakeCase } from '~/lib/utils/text_utility';
import { hideFlash } from '~/flash';
import {
  newDate,
  dayAfter,
  secondsToDays,
  getDatesInRange,
  getDayDifference,
  getDateInPast,
  getDateInFuture,
} from '~/lib/utils/datetime_utility';
import { dateFormats } from '../shared/constants';
import { STAGE_NAME } from './constants';
import { toYmd } from '../shared/utils';

const EVENT_TYPE_LABEL = 'label';

export const removeFlash = (type = 'alert') => {
  const flashEl = document.querySelector(`.flash-${type}`);
  if (flashEl) {
    hideFlash(flashEl);
  }
};

export const toggleSelectedLabel = ({ selectedLabelIds = [], value = null }) => {
  if (!value) return selectedLabelIds;
  return selectedLabelIds.includes(value)
    ? selectedLabelIds.filter(v => v !== value)
    : [...selectedLabelIds, value];
};

export const isStartEvent = ev => Boolean(ev) && Boolean(ev.canBeStartEvent) && ev.canBeStartEvent;

export const eventToOption = (obj = null) => {
  if (!obj || (!obj.text && !obj.identifier)) return null;
  const { name: text = '', identifier: value = null } = obj;
  return { text, value };
};

export const getAllowedEndEvents = (events = [], targetIdentifier = null) => {
  if (!targetIdentifier || !events.length) return [];
  const st = events.find(({ identifier }) => identifier === targetIdentifier);
  return st && st.allowedEndEvents ? st.allowedEndEvents : [];
};

export const eventsByIdentifier = (events = [], targetIdentifier = []) => {
  if (!targetIdentifier || !targetIdentifier.length || !events.length) return [];
  return events.filter(({ identifier = '' }) => targetIdentifier.includes(identifier));
};

export const isLabelEvent = (labelEvents = [], ev = null) =>
  Boolean(ev) && labelEvents.length && labelEvents.includes(ev);

export const getLabelEventsIdentifiers = (events = []) =>
  events.filter(ev => ev.type && ev.type === EVENT_TYPE_LABEL).map(i => i.identifier);

/**
 * Checks if the specified stage is in memory or persisted to storage based on the id
 *
 * Default value stream analytics stages are initially stored in memory, when they are first
 * created the id for the stage is the name of the stage in lowercase. This string id
 * is used to fetch stage data (events, median calculation)
 *
 * When either a custom stage is created or an edit is made to a default stage then the
 * default stages get persisted to storage and will have a numeric id. The new numeric
 * id should then be used to access stage data
 *
 */
export const isPersistedStage = ({ custom, id }) => custom || isNumber(id);

/**
 * Returns the the correct slug to use for a stage
 * default stages use the snakecased title of the stage, while custom
 * stages will have a numeric id
 *
 * @param {Object} obj
 * @param {string} obj.title - title of the stage
 * @param {number} obj.id - numerical object id available for custom stages
 * @param {boolean} obj.custom - boolean flag indicating a custom stage
 * @returns {(number|string)} Returns a numerical id for customs stages and string for default stages
 */
const stageUrlSlug = ({ id, title, custom = false }) => {
  if (custom) return id;
  // We still use 'production' as the id to access this stage, even though the title is 'Total'
  return title.toLowerCase() === STAGE_NAME.TOTAL
    ? STAGE_NAME.PRODUCTION
    : convertToSnakeCase(title);
};

export const transformRawStages = (stages = []) =>
  stages.map(({ id, title, name = '', custom = false, ...rest }) => ({
    ...convertObjectPropsToCamelCase(rest, { deep: true }),
    id,
    title,
    custom,
    slug: isPersistedStage({ custom, id }) ? id : stageUrlSlug({ custom, id, title }),
    // the name field is used to create a stage, but the get request returns title
    name: name.length ? name : title,
  }));

export const transformRawTasksByTypeData = (data = []) => {
  if (!data.length) return [];
  return data.map(d => convertObjectPropsToCamelCase(d, { deep: true }));
};

/**
 * Takes the duration data for selected stages, transforms the date values and returns
 * the data in a flattened array
 *
 * The received data is expected to be the following format; One top level object in the array per stage,
 * each potentially having multiple data entries.
 * [
 *   {
 *    slug: 'issue',
 *    selected: true,
 *    data: [
 *      {
 *        'duration_in_seconds': 1234,
 *        'finished_at': '2019-09-02T18:25:43.511Z'
 *      },
 *      ...
 *    ]
 *   },
 *   ...
 * ]
 *
 * The data is then transformed and flattened into the following format;
 * [
 *  {
 *    'duration_in_seconds': 1234,
 *    'finished_at': '2019-09-02'
 *  },
 *  ...
 * ]
 *
 * @param {Array} data - The duration data for selected stages
 * @returns {Array} An array with each item being an object containing the duration_in_seconds and finished_at values for an event
 */
export const flattenDurationChartData = data =>
  data
    .map(stage =>
      stage.data.map(event => {
        const date = new Date(event.finished_at);
        return {
          ...event,
          finished_at: dateFormat(date, dateFormats.isoDate),
        };
      }),
    )
    .flat();

/**
 * Takes the duration data for selected stages, groups the data by day and calculates the total duration
 * per day.
 *
 * The received data is expected to be the following format; One top level object in the array per stage,
 * each potentially having multiple data entries.
 * [
 *   {
 *    slug: 'issue',
 *    selected: true,
 *    data: [
 *      {
 *        'duration_in_seconds': 1234,
 *        'finished_at': '2019-09-02T18:25:43.511Z'
 *      },
 *      ...
 *    ]
 *   },
 *   ...
 * ]
 *
 * The data is then computed and transformed into a format that can be passed to the chart:
 * [
 *  ['2019-09-02', 7, '2019-09-02'],
 *  ['2019-09-03', 10, '2019-09-03'],
 *  ['2019-09-04', 8, '2019-09-04'],
 *  ...
 * ]
 *
 * In the data above, each array i represents a point in the scatterplot with the following data:
 * i[0] = date, displayed on x axis
 * i[1] = metric, displayed on y axis
 * i[2] = date, used in the tooltip
 *
 * @param {Array} data - The duration data for selected stages
 * @param {Date} startDate - The globally selected cycle analytics start date
 * @param {Date} endDate - The globally selected cycle analytics end date
 * @returns {Array} An array with each item being another arry of three items (plottable date, computed total, tooltip display date)
 */
export const getDurationChartData = (data, startDate, endDate) => {
  const flattenedData = flattenDurationChartData(data);
  const eventData = [];

  for (
    let currentDate = newDate(startDate);
    currentDate <= endDate;
    currentDate = dayAfter(currentDate)
  ) {
    const currentISODate = dateFormat(newDate(currentDate), dateFormats.isoDate);
    const valuesForDay = flattenedData.filter(object => object.finished_at === currentISODate);
    const summedData = valuesForDay.reduce((total, value) => total + value.duration_in_seconds, 0);
    const summedDataInDays = secondsToDays(summedData);

    if (summedDataInDays) eventData.push([currentISODate, summedDataInDays, currentISODate]);
  }

  return eventData;
};

/**
 * Takes the offset duration data for selected stages and calls getDurationChartData to compute the totals.
 * The data is then transformed into a format expected by the scatterplot;
 *
 * [
 *   ['2019-09-02', 7],
 *   ...
 * ]
 *
 * The transformation works by calling getDateInPast on the provided startDate and endDate in order to match
 * the startDate and endDate fetched when making the API call to fetch the data.
 *
 * In order to map the offset data to plottable points within the chart's range, getDateInFuture is called
 * on the data series with the same offest used for getDateInPast. This creates plottable data that matches up
 * with the data being displayed on the chart.
 *
 * @param {Array} data - The computed, plottable duration chart data
 * @param {Date} startDate - The globally selected cycle analytics start date
 * @param {Date} endDate - The globally selected cycle analytics end date
 * @returns {Array} An array with each item being another arry of two items (date, computed total)
 */
export const getDurationChartMedianData = (data, startDate, endDate) => {
  const offsetValue = getDayDifference(startDate, endDate);
  const offsetEndDate = getDateInPast(endDate, offsetValue);
  const offsetStartDate = getDateInPast(startDate, offsetValue);

  const offsetDurationData = getDurationChartData(data, offsetStartDate, offsetEndDate);

  const result = offsetDurationData.map(event => [
    dateFormat(getDateInFuture(new Date(event[0]), offsetValue), dateFormats.isoDate),
    event[1],
  ]);

  return result;
};

export const orderByDate = (a, b, dateFmt = datetime => new Date(datetime).getTime()) =>
  dateFmt(a) - dateFmt(b);

/**
 * Takes a dictionary of dates and the associated value, sorts them and returns just the value
 *
 * @param {Object.<Date, number>} series - Key value pair of dates and the value for that date
 * @returns {number[]} The values of each key value pair
 */
export const flattenTaskByTypeSeries = (series = {}) =>
  Object.entries(series)
    .sort((a, b) => orderByDate(a[0], b[0]))
    .map(dataSet => dataSet[1]);

/**
 * @typedef {Object} RawTasksByTypeData
 * @property {Object} label - Raw data for a group label
 * @property {Array} series - Array of arrays with date and associated value ie [ ['2020-01-01', 10],['2020-01-02', 10] ]

 * @typedef {Object} TransformedTasksByTypeData
 * @property {Array} groupBy - The list of dates for the range of data in each data series
 * @property {Array} data - An array of the data values for each series
 * @property {Array} seriesNames - Names of the series to be charted ie label names
 */

/**
 * Takes the raw tasks by type data and generates an array of data points,
 * an array of data series and an array of data labels for the given time period.
 *
 * Currently the data is transformed to support use in a stacked column chart:
 * https://gitlab-org.gitlab.io/gitlab-ui/?path=/story/charts-stacked-column-chart--stacked
 *
 * @param {Object} obj
 * @param {RawTasksByTypeData[]} obj.data - array of raw data, each element contains a label and series
 * @param {Date} obj.startDate - start date in ISO date format
 * @param {Date} obj.endDate - end date in ISO date format
 *
 * @returns {TransformedTasksByTypeData} The transformed data ready for use in charts
 */
export const getTasksByTypeData = ({ data = [], startDate = null, endDate = null }) => {
  if (!startDate || !endDate || !data.length) {
    return {
      groupBy: [],
      data: [],
      seriesNames: [],
    };
  }

  const groupBy = getDatesInRange(startDate, endDate, toYmd).sort(orderByDate);
  const zeroValuesForEachDataPoint = groupBy.reduce(
    (acc, date) => ({
      ...acc,
      [date]: 0,
    }),
    {},
  );

  const transformed = data.reduce(
    (acc, curr) => {
      const {
        label: { title },
        series,
      } = curr;
      acc.seriesNames = [...acc.seriesNames, title];
      acc.data = [
        ...acc.data,
        // adds 0 values for each data point and overrides with data from the series
        flattenTaskByTypeSeries({ ...zeroValuesForEachDataPoint, ...Object.fromEntries(series) }),
      ];
      return acc;
    },
    {
      data: [],
      seriesNames: [],
    },
  );

  return {
    ...transformed,
    groupBy,
  };
};
