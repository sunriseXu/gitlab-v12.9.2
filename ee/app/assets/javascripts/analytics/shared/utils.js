import dateFormat from 'dateformat';
import { dateFormats } from './constants';
import { convertObjectPropsToCamelCase } from '~/lib/utils/common_utils';

export const toYmd = date => dateFormat(date, dateFormats.isoDate);

export default {
  toYmd,
};

export const formattedDate = d => dateFormat(d, dateFormats.defaultDate);

/**
 * Creates a group object from a dataset. Returns null if no groupId is present.
 *
 * @param {Object} dataset - The container's dataset
 * @returns {Object} - A group object
 */
export const buildGroupFromDataset = dataset => {
  const { groupId, groupName, groupFullPath, groupAvatarUrl } = dataset;

  if (groupId) {
    return {
      id: Number(groupId),
      name: groupName,
      full_path: groupFullPath,
      avatar_url: groupAvatarUrl,
    };
  }

  return null;
};

/**
 * Creates a project object from a dataset. Returns null if no projectId is present.
 *
 * @param {Object} dataset - The container's dataset
 * @returns {Object} - A project object
 */
export const buildProjectFromDataset = dataset => {
  const { projectId, projectName, projectPathWithNamespace, projectAvatarUrl } = dataset;

  if (projectId) {
    return {
      id: Number(projectId),
      name: projectName,
      path_with_namespace: projectPathWithNamespace,
      avatar_url: projectAvatarUrl,
    };
  }

  return null;
};

/**
 * Creates an array of project objects from a json string. Returns null if no projects are present.
 *
 * @param {String} data - JSON encoded array of projects
 * @returns {Array} - An array of project objects
 */
const buildProjectsFromJSON = (projects = []) => {
  if (!projects.length) return [];
  return JSON.parse(projects);
};

/**
 * Builds the initial data object for cycle analytics with data loaded from the backend
 *
 * @param {Object} dataset - dataset object paseed to the frontend via data-* properties
 * @returns {Object} - The initial data to load the app with
 */
export const buildCycleAnalyticsInitialData = ({
  groupId = null,
  createdBefore = null,
  createdAfter = null,
  projects = null,
  groupName = null,
  groupFullPath = null,
  groupAvatarUrl = null,
} = {}) => ({
  group: groupId
    ? convertObjectPropsToCamelCase(
        buildGroupFromDataset({ groupId, groupName, groupFullPath, groupAvatarUrl }),
      )
    : null,
  createdBefore: createdBefore ? new Date(createdBefore) : null,
  createdAfter: createdAfter ? new Date(createdAfter) : null,
  selectedProjects: projects
    ? buildProjectsFromJSON(projects).map(convertObjectPropsToCamelCase)
    : [],
});
