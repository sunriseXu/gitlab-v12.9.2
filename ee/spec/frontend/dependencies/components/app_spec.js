import { GlBadge, GlEmptyState, GlLoadingIcon, GlTab, GlLink } from '@gitlab/ui';
import { mount } from '@vue/test-utils';
import { TEST_HOST } from 'helpers/test_constants';
import createStore from 'ee/dependencies/store';
import { addListType } from 'ee/dependencies/store/utils';
import { DEPENDENCY_LIST_TYPES } from 'ee/dependencies/store/constants';
import { REPORT_STATUS } from 'ee/dependencies/store/modules/list/constants';
import DependenciesApp from 'ee/dependencies/components/app.vue';
import DependenciesActions from 'ee/dependencies/components/dependencies_actions.vue';
import DependencyListIncompleteAlert from 'ee/dependencies/components/dependency_list_incomplete_alert.vue';
import DependencyListJobFailedAlert from 'ee/dependencies/components/dependency_list_job_failed_alert.vue';
import PaginatedDependenciesTable from 'ee/dependencies/components/paginated_dependencies_table.vue';
import { getDateInPast } from '~/lib/utils/datetime_utility';

describe('DependenciesApp component', () => {
  let store;
  let wrapper;
  const { namespace: allNamespace } = DEPENDENCY_LIST_TYPES.all;
  const { namespace: vulnerableNamespace } = DEPENDENCY_LIST_TYPES.vulnerable;

  const basicAppProps = {
    endpoint: '/foo',
    emptyStateSvgPath: '/bar.svg',
    documentationPath: TEST_HOST,
    supportDocumentationPath: `${TEST_HOST}/dependency_scanning#supported-languages`,
  };

  const factory = (props = basicAppProps) => {
    store = createStore();
    addListType(store, DEPENDENCY_LIST_TYPES.vulnerable);
    jest.spyOn(store, 'dispatch').mockImplementation();

    const canBeStubbed = component => !['GlTab', 'GlTabs'].includes(component);
    const stubs = Object.keys(DependenciesApp.components).filter(canBeStubbed);

    wrapper = mount(DependenciesApp, {
      store,
      propsData: { ...props },
      stubs,
    });
  };

  const setStateJobNotRun = () => {
    Object.assign(store.state[allNamespace], {
      initialized: true,
      isLoading: false,
      dependencies: [],
    });
    store.state[allNamespace].pageInfo.total = 0;
    store.state[allNamespace].reportInfo.status = REPORT_STATUS.jobNotSetUp;
  };

  const setStateLoaded = () => {
    [allNamespace, vulnerableNamespace].forEach((namespace, i, { length }) => {
      const total = length - i;
      Object.assign(store.state[namespace], {
        initialized: true,
        isLoading: false,
        dependencies: Array(total)
          .fill(null)
          .map((_, id) => ({ id })),
      });
      store.state[namespace].pageInfo.total = total;
      store.state[namespace].reportInfo.status = REPORT_STATUS.ok;
      store.state[namespace].reportInfo.generatedAt = getDateInPast(new Date(), 7);
      store.state[namespace].reportInfo.jobPath = '/jobs/foo/321';
    });
  };

  const setStateJobFailed = () => {
    Object.assign(store.state[allNamespace], {
      initialized: true,
      isLoading: false,
      dependencies: [],
    });
    store.state[allNamespace].pageInfo.total = 0;
    store.state[allNamespace].reportInfo.status = REPORT_STATUS.jobFailed;
    store.state[allNamespace].reportInfo.jobPath = '/jobs/foo/321';
  };

  const setStateListIncomplete = () => {
    Object.assign(store.state[allNamespace], {
      initialized: true,
      isLoading: false,
      dependencies: [{ id: 0 }],
    });
    store.state[allNamespace].pageInfo.total = 1;
    store.state[allNamespace].reportInfo.status = REPORT_STATUS.incomplete;
  };

  const setStateNoDependencies = () => {
    Object.assign(store.state[allNamespace], {
      initialized: true,
      isLoading: false,
      dependencies: [],
    });
    store.state[allNamespace].pageInfo.total = 0;
    store.state[allNamespace].reportInfo.status = REPORT_STATUS.noDependencies;
  };

  const findJobFailedAlert = () => wrapper.find(DependencyListJobFailedAlert);
  const findIncompleteListAlert = () => wrapper.find(DependencyListIncompleteAlert);
  const findDependenciesTables = () => wrapper.findAll(PaginatedDependenciesTable);
  const findTabControls = () => wrapper.findAll('.gl-tab-nav-item');
  const findVulnerableTabControl = () => findTabControls().at(1);
  const findVulnerableTabComponent = () => wrapper.findAll(GlTab).at(1);

  const findHeader = () => wrapper.find('section > header');
  const findHeaderHelpLink = () => findHeader().find(GlLink);
  const findHeaderJobLink = () => findHeader().find('a');

  const expectComponentWithProps = (Component, props = {}) => {
    const componentWrapper = wrapper.find(Component);
    expect(componentWrapper.isVisible()).toBe(true);
    expect(componentWrapper.props()).toEqual(expect.objectContaining(props));
  };

  const expectComponentPropsToMatchSnapshot = Component => {
    const componentWrapper = wrapper.find(Component);
    expect(componentWrapper.props()).toMatchSnapshot();
  };

  const expectNoDependenciesTables = () => expect(findDependenciesTables()).toHaveLength(0);
  const expectNoHeader = () => expect(findHeader().exists()).toBe(false);

  const expectDependenciesTables = () => {
    const tables = findDependenciesTables();
    expect(tables).toHaveLength(2);
    expect(tables.at(0).props()).toEqual({ namespace: allNamespace });
    expect(tables.at(1).props()).toEqual({ namespace: vulnerableNamespace });
  };

  const expectHeader = () => {
    expect(findHeader().exists()).toBe(true);
  };

  afterEach(() => {
    wrapper.destroy();
  });

  describe('on creation', () => {
    beforeEach(() => {
      factory();
    });

    it('dispatches the correct initial actions', () => {
      expect(store.dispatch.mock.calls).toEqual([
        ['setDependenciesEndpoint', basicAppProps.endpoint],
        ['fetchDependencies'],
      ]);
    });

    it('shows only the loading icon', () => {
      expectComponentWithProps(GlLoadingIcon);
      expectNoHeader();
      expectNoDependenciesTables();
    });

    describe('given the dependency list job has not yet run', () => {
      beforeEach(() => {
        setStateJobNotRun();

        return wrapper.vm.$nextTick();
      });

      it('shows only the empty state', () => {
        expectComponentWithProps(GlEmptyState, { svgPath: basicAppProps.emptyStateSvgPath });
        expectComponentPropsToMatchSnapshot(GlEmptyState);
        expectNoHeader();
        expectNoDependenciesTables();
      });
    });

    describe('given a list of dependencies and ok report', () => {
      beforeEach(() => {
        setStateLoaded();

        return wrapper.vm.$nextTick();
      });

      it('shows both dependencies tables with the correct props', () => {
        expectHeader();
        expectDependenciesTables();
      });

      it('shows a link to the latest job', () => {
        expect(findHeaderJobLink().attributes('href')).toBe('/jobs/foo/321');
      });

      it('shows when the last job ran', () => {
        expect(findHeader().text()).toContain('1 week ago');
      });

      it('shows a link to the dependencies documentation page', () => {
        expect(findHeaderHelpLink().attributes('href')).toBe(TEST_HOST);
      });

      it('displays the tabs correctly', () => {
        const expected = [
          {
            text: 'All',
            total: '2',
          },
          {
            text: 'Vulnerable',
            total: '1',
          },
        ];

        const tabs = findTabControls();
        expected.forEach(({ text, total }, i) => {
          const tab = tabs.at(i);
          expect(tab.text()).toEqual(expect.stringContaining(text));
          expect(
            tab
              .find(GlBadge)
              .text()
              .trim(),
          ).toEqual(total);
        });
      });

      it('passes the correct namespace to dependencies actions component', () => {
        expectComponentWithProps(DependenciesActions, { namespace: allNamespace });
      });

      describe('given the user clicks on the vulnerable tab', () => {
        beforeEach(() => {
          findVulnerableTabControl().trigger('click');

          return wrapper.vm.$nextTick();
        });

        it('changes the current list', () => {
          expect(store.dispatch).toHaveBeenCalledWith('setCurrentList', vulnerableNamespace);
        });
      });

      describe('given the current list is the vulnerable dependencies list', () => {
        const namespace = vulnerableNamespace;
        beforeEach(() => {
          store.state.currentList = namespace;

          return wrapper.vm.$nextTick();
        });

        it('passes the correct namespace to dependencies actions component', () => {
          expectComponentWithProps(DependenciesActions, { namespace });
        });
      });

      it('has enabled vulnerable tab', () => {
        expect(findVulnerableTabComponent().classes('disabled')).toBe(false);
      });

      describe('given there are no vulnerable dependencies', () => {
        beforeEach(() => {
          store.state[vulnerableNamespace].dependencies = [];
          store.state[vulnerableNamespace].pageInfo.total = 0;

          return wrapper.vm.$nextTick();
        });

        it('disables the vulnerable tab', () => {
          expect(findVulnerableTabComponent().classes('disabled')).toBe(true);
        });
      });

      describe('given the user has public permissions', () => {
        beforeEach(() => {
          store.state[allNamespace].reportInfo.generatedAt = '';
          store.state[allNamespace].reportInfo.jobPath = '';

          return wrapper.vm.$nextTick();
        });

        it('shows the header', () => {
          expectHeader();
        });

        it('does not show when the last job ran', () => {
          expect(findHeader().text()).not.toContain('1 week ago');
        });

        it('does not show a link to the latest job', () => {
          expect(findHeaderJobLink().exists()).toBe(false);
        });
      });
    });

    describe('given the dependency list job failed', () => {
      beforeEach(() => {
        setStateJobFailed();

        return wrapper.vm.$nextTick();
      });

      it('passes the correct props to the job failure alert', () => {
        expectComponentWithProps(DependencyListJobFailedAlert, {
          jobPath: '/jobs/foo/321',
        });
      });

      it('shows both dependencies tables with the correct props', expectDependenciesTables);

      describe('when the job failure alert emits the dismiss event', () => {
        beforeEach(() => {
          const alertWrapper = findJobFailedAlert();
          alertWrapper.vm.$emit('dismiss');
          return wrapper.vm.$nextTick();
        });

        it('does not render the job failure alert', () => {
          expect(findJobFailedAlert().exists()).toBe(false);
        });
      });
    });

    describe('given a dependency list which is known to be incomplete', () => {
      beforeEach(() => {
        setStateListIncomplete();

        return wrapper.vm.$nextTick();
      });

      it('passes the correct props to the incomplete-list alert', () => {
        expectComponentWithProps(DependencyListIncompleteAlert);
      });

      it('shows both dependencies tables with the correct props', expectDependenciesTables);

      describe('when the incomplete-list alert emits the dismiss event', () => {
        beforeEach(() => {
          const alertWrapper = findIncompleteListAlert();
          alertWrapper.vm.$emit('dismiss');
          return wrapper.vm.$nextTick();
        });

        it('does not render the incomplete-list alert', () => {
          expect(findIncompleteListAlert().exists()).toBe(false);
        });
      });
    });

    describe('given there are no dependencies detected', () => {
      beforeEach(() => {
        setStateNoDependencies();
      });

      it('shows only the empty state', () => {
        expectComponentWithProps(GlEmptyState, { svgPath: basicAppProps.emptyStateSvgPath });
        expectComponentPropsToMatchSnapshot(GlEmptyState);
        expectNoHeader();
        expectNoDependenciesTables();
      });
    });
  });
});
