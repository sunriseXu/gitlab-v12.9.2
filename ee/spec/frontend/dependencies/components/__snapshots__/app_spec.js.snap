// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DependenciesApp component on creation given the dependency list job has not yet run shows only the empty state 1`] = `
Object {
  "compact": false,
  "description": "The dependency list details information about the components used within your project.",
  "primaryButtonLink": null,
  "primaryButtonText": null,
  "secondaryButtonLink": null,
  "secondaryButtonText": null,
  "svgPath": "/bar.svg",
  "title": "View dependency details for your project",
}
`;

exports[`DependenciesApp component on creation given there are no dependencies detected shows only the empty state 1`] = `
Object {
  "compact": false,
  "description": "It seems like the Dependency Scanning job ran successfully, but no dependencies have been detected in your project.",
  "primaryButtonLink": null,
  "primaryButtonText": null,
  "secondaryButtonLink": null,
  "secondaryButtonText": null,
  "svgPath": "/bar.svg",
  "title": "Dependency List has no entries",
}
`;
